### ISSUE TITLE
Crash on shutdown while performing maintenance

### ISSUE THREAD

**Issue by marta-lokhova** [stellar/stellar-core#3106]
Needs to be investigated. The crash occurred when stellar-core was restarted (as part of v17.2.0rc2 -> v.17.2.0 upgrade):
```
2021-06-30T00:03:36.491 GCW5Q [Tx INFO] applying ledger 223875 (txs:2, ops:6)
2021-06-30T00:03:36.505 GCW5Q [Ledger INFO] Closed ledger: [seq=223875, hash=5231b0]
2021-06-30T00:03:40.193 GCW5Q [default INFO] got signal 15, shutting down
2021-06-30T00:03:40.526 GCW5Q [default INFO] Performing maintenance
2021-06-30T00:03:40.526 GCW5Q [History INFO] Trimming history <= ledger 150920 (rmin=150920, qmin=223875, lmin=223811)
2021-06-30T00:03:41.405 GBH7K [default INFO] Starting stellar-core stellar-core 17.2.0 (e47d483433ce4f7eaf0a0512e2556beb482c8851)
2021-06-30T00:03:41.407 GBH7K [Database INFO] Connecting to: postgresql://dbname=stellar user=stellar password=******** host=/var/run/postgresql
2021-06-30T00:03:41.412 GBH7K [SCP INFO] LocalNode::LocalNode@GBH7K qSet: 59d361
2021-06-30T00:03:41.414 GBH7K [Database INFO] DB schema is in current version
```

Note that maintenance ran _after_ core received the shutdown signal, so there might be an issue related to object lifetime during shutdown

**Comment by marta-lokhova** [comment]
stack trace from the core dump. Looks like the logs are a bit misleading in this case, as the crash happened in Work
```
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007f944f841859 in __GI_abort () at abort.c:79
#2  0x00007f944f841729 in __assert_fail_base (fmt=0x7f944f9d7588 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n",
    assertion=0xa92348 "!hasChildren()", file=0xa92357 "work/Work.cpp", line=21, function=<optimized out>)
    at assert.c:92
#3  0x00007f944f852f36 in __GI___assert_fail (assertion=0xa92348 "!hasChildren()",
    file=0xa92357 "work/Work.cpp", line=21, function=0xa92365 "virtual stellar::Work::~Work()") at assert.c:101
#4  0x00000000007f1883 in stellar::Work::~Work (this=<optimized out>) at work/Work.cpp:21
#5  0x00000000007f35e4 in stellar::WorkScheduler::~WorkScheduler (this=0x2475750) at work/WorkScheduler.cpp:23
#6  stellar::WorkScheduler::~WorkScheduler (this=0x2475750) at work/WorkScheduler.cpp:22
#7  0x000000000064ef79 in std::__1::__shared_count::__release_shared (this=0x24e12a0)
    at /usr/lib/llvm-8/bin/../include/c++/v1/memory:3544
#8  std::__1::__shared_weak_count::__release_shared (this=0x24e12a0)
    at /usr/lib/llvm-8/bin/../include/c++/v1/memory:3586
#9  std::__1::shared_ptr<stellar::WorkScheduler>::~shared_ptr (this=<optimized out>)
    at /usr/lib/llvm-8/bin/../include/c++/v1/memory:4522
#10 stellar::ApplicationImpl::~ApplicationImpl (this=0x246df18) at main/ApplicationImpl.cpp:573
#11 0x00000000006a14fe in std::__1::__shared_count::__release_shared (this=0x246df00)
    at /usr/lib/llvm-8/bin/../include/c++/v1/memory:3544
#12 std::__1::__shared_weak_count::__release_shared (this=0x246df00)
    at /usr/lib/llvm-8/bin/../include/c++/v1/memory:3586
#13 std::__1::shared_ptr<stellar::Application>::~shared_ptr (this=<optimized out>)
    at /usr/lib/llvm-8/bin/../include/c++/v1/memory:4522
#14 stellar::run(stellar::CommandLineArgs const&)::$_3::operator()() const (this=<optimized out>)
    at main/CommandLine.cpp:1191
#15 std::__1::__invoke<stellar::run(stellar::CommandLineArgs const&)::$_3&> (__f=...)
    at /usr/lib/llvm-8/bin/../include/c++/v1/type_traits:4353
#16 std::__1::__invoke_void_return_wrapper<int>::__call<stellar::run(stellar::CommandLineArgs const&)::$_3&>(stellar::run(stellar::CommandLineArgs const&)::$_3&) (__args=...)
    at /usr/lib/llvm-8/bin/../include/c++/v1/__functional_base:318
#17 std::__1::__function::__alloc_func<stellar::run(stellar::CommandLineArgs const&)::$_3, std::__1::allocator<stellar::run(stellar::CommandLineArgs const&)::$_3>, int ()>::operator()() (this=<optimized out>)
    at /usr/lib/llvm-8/bin/../include/c++/v1/functional:1527
#18 std::__1::__function::__func<stellar::run(stellar::CommandLineArgs const&)::$_3, std::__1::allocator<stellar::run(stellar::CommandLineArgs const&)::$_3>, int ()>::operator()() (this=<optimized out>)
    at /usr/lib/llvm-8/bin/../include/c++/v1/functional:1651
#19 0x000000000067cfbd in std::__1::__function::__value_func<int ()>::operator()() const (this=<optimized out>)
    at /usr/lib/llvm-8/bin/../include/c++/v1/functional:1799
#20 std::__1::function<int ()>::operator()() const (this=<optimized out>)
    at /usr/lib/llvm-8/bin/../include/c++/v1/functional:2347
#21 stellar::(anonymous namespace)::runWithHelp(stellar::CommandLineArgs const&, std::__1::vector<stellar::(anonymous namespace)::ParserWithValidation, std::__1::allocator<stellar::(anonymous namespace)::ParserWithValidation> >, std::__1::function<int ()>) (args=..., parsers=..., f=...) at main/CommandLine.cpp:409
#22 0x000000000068dea9 in stellar::run (args=...) at main/CommandLine.cpp:1098
#23 0x0000000000694675 in std::__1::__function::__value_func<int (stellar::CommandLineArgs const&)>::operator()(stellar::CommandLineArgs const&) const (this=<optimized out>, __args=...)
    at /usr/lib/llvm-8/bin/../include/c++/v1/functional:1799
#24 std::__1::function<int (stellar::CommandLineArgs const&)>::operator()(stellar::CommandLineArgs const&) const
    (this=<optimized out>, __arg=...) at /usr/lib/llvm-8/bin/../include/c++/v1/functional:2347
#25 stellar::(anonymous namespace)::CommandLine::Command::run (this=<optimized out>, args=...)
    at main/CommandLine.cpp:450
#26 stellar::handleCommandLine (argc=4, argv=0x7ffd808af718) at main/CommandLine.cpp:1630
#27 0x00000000006f75b4 in main (argc=4, argv=0x7ffd808af718) at main/main.cpp:55
```

**Comment by hidenori-shinohara** [comment]
tl; dr --- `WorkScheduler` may continue to schedule more work after it's aborted. (It doesn't while it's "aborting", but it does after it's "aborted"). It can be prevented by updating `scheduleWork`.


---

The longer version

The stack trace that @marta-lokhova posted above shows that the issue is actually inside the work class.

More specifically, the following `assert` fails:

```
Work::~Work()
{
    // Work is destroyed only if in terminal state, and is properly reset
    assert(!hasChildren());
}
```

# Why this "usually" doesn't happen

Both `ApplicationImpl::gracefulStop()` and `~ApplicationImpl()` call `ApplicationImpl::shutdownWorkScheduler()` which calls `WorkScheduler::shutdown()`.

`WorkScheduler::shutdown()` first calls `Work::shutdown()` which calls `BasicWork::shutdown()` which sets the state to `ABORTING`.

Then `WorkScheduler::shutdown()` calls `scheduleOne()` on itself, which continues to call `crankWork()` until the state is `ABORTED`.

In order to set the state to `ABORTED`, `BasicWork::crankWork()` calls `BasicWork::setState(ABORTED)` which calls `BasicWork::reset()` which calls `Work::onReset()` which calls `Work::clearChildren()`.

# Why this happened this time

`ApplicationImpl::shutdownWorkScheduler()` correctly clears `mChildren`, but the issue is what happens AFTER that.

See the implementation of `scheduleWork`

```
    // Returns a work that's been scheduled, or nullptr if the WorkScheduler
    // is aborting.
    template <typename T, typename... Args>
    std::shared_ptr<T>
    scheduleWork(Args&&... args)
    {
        if (isAborting())
        {
            return nullptr;
        }
...
```
The `scheduleWork` function only checks if it's aborting, and schedules stuff if it's already been `ABORTED`. Thus if somehow we try to schedule work after it's been aborted, then we can still do so and end up with a non-empty `mChildren` upon the destruction call.

In fact, if you add something like
```
std::vector<std::shared_ptr<BasicWork>> seq;                                                                                                                                                                                        
        std::shared_ptr<BasicWork> ptr =                                                                                                                                                                                                    
            getWorkScheduler().scheduleWork<WorkSequence>("work-sequence", seq);
```
inside `~ApplicationImpl` AFTER `shutdownWorkScheduler();` has been called, then the code fails the assertion `!hasChildren()`.

This can be fixed by updating `if (isAborting())` to `if (isAborting() || isDone())`.

# Another observation

The command that I used to reproduce this issue was
```
ps ax  | grep "stellar-core run" | head -n 1 | awk '{print $1;}' | xargs kill -15 && \
    stellar-core http-command --conf docs/stellar-core_example.cfg "maintenance?queue=true"
```

This leads to Core running maintenance after the signal.

```
2021-08-05T13:15:22.206 GBA2J [default INFO] got signal 15, shutting down
2021-08-05T13:15:22.215 GBA2J [default INFO] Performing maintenance
```

This happens because `CommandHandler` continues to process http commands after signal 15 has been received.
I'm not sure if this is an issue. I don't really see how this is "wrong". If it is an issue, it seems fairly insignificant. This is encountered only if one decides to run an http command after Core decided to shut down and before it actually shuts down which I believe is generally less than a few seconds.

**Comment by hidenori-shinohara** [comment]
Thank you @marta for the repro!

I was able to reproduce the issue locally using the same approach, and the change proposed above indeed fixed it as well.

(inlined some comments below)

```
^C2021-08-09T13:21:52.873 GB7TV [default INFO] got signal 2, shutting down <---- This is when I hit Ctrl+C
2021-08-09T13:21:52.873 GB7TV [Work TRACE] Shutting down: wait-for-lcl
2021-08-09T13:21:52.873 GB7TV [Work DEBUG] work wait-for-lcl : WORK_WAITING -> WORK_ABORTING
2021-08-09T13:21:52.873 GB7TV [Work TRACE] Shutting down: self-check
2021-08-09T13:21:52.873 GB7TV [Work DEBUG] work self-check : WORK_WAITING -> WORK_ABORTING
2021-08-09T13:21:52.873 GB7TV [Work TRACE] Shutting down: work-scheduler
2021-08-09T13:21:52.873 GB7TV [Work DEBUG] work work-scheduler : WORK_WAITING -> WORK_ABORTING
2021-08-09T13:21:52.873 GB7TV [Work TRACE] Abort progress for wait-for-lcl: done
2021-08-09T13:21:52.874 GB7TV [Work TRACE] resetting wait-for-lcl
2021-08-09T13:21:52.874 GB7TV [Work DEBUG] work wait-for-lcl : WORK_ABORTING -> WORK_ABORTED
2021-08-09T13:21:52.874 GB7TV [Work TRACE] Abort progress for self-check: still aborting
2021-08-09T13:21:52.874 GB7TV [Work TRACE] Abort progress for work-scheduler: still aborting
2021-08-09T13:21:52.874 GB7TV [Work TRACE] work-scheduler: waiting for children to abort.
2021-08-09T13:21:52.874 GB7TV [Work TRACE] Abort progress for work-scheduler: still aborting
2021-08-09T13:21:52.874 GB7TV [Work TRACE] Abort progress for self-check: done
2021-08-09T13:21:52.874 GB7TV [Work TRACE] resetting self-check
2021-08-09T13:21:52.874 GB7TV [Work DEBUG] work self-check : WORK_ABORTING -> WORK_ABORTED
2021-08-09T13:21:52.874 GB7TV [Work TRACE] Abort progress for work-scheduler: still aborting
2021-08-09T13:21:52.874 GB7TV [Work TRACE] work-scheduler: waiting for children to abort.
2021-08-09T13:21:52.874 GB7TV [Work TRACE] Abort progress for work-scheduler: done
2021-08-09T13:21:52.874 GB7TV [Work TRACE] resetting work-scheduler
2021-08-09T13:21:52.874 GB7TV [Work DEBUG] work work-scheduler : WORK_ABORTING -> WORK_ABORTED <--- WorkScheduler has been aborted
2021-08-09T13:21:52.874 GB7TV [Bucket TRACE] b72bc26d469a3f5079510767e2ea4773e02da574dec72e2f32f5828b79e0af40 referenced by bucket list
2021-08-09T13:21:52.874 GB7TV [Bucket TRACE] 0000000000000000000000000000000000000000000000000000000000000000 referenced by bucket list
2021-08-09T13:21:53.178 GB7TV [default INFO] ApplicationImpl::scheduleSelfCheck called <--- We still try to schedule self-check here.
2021-08-09T13:21:53.178 GB7TV [Work INFO] scheduleWork refuses to schedule work since isDone() == true <--- The change proposed above ensures that we won't schedule anything once isDone() == true.
2021-08-09T13:21:53.874 GB7TV [default INFO] Application destructing
2021-08-09T13:21:53.874 GB7TV [Bucket WARNING] BucketManager: shutdown is called more than once
2021-08-09T13:21:53.874 GB7TV [default DEBUG] Joining 11 worker threads
2021-08-09T13:21:53.874 GB7TV [default DEBUG] Joined all 11 threads
2021-08-09T13:21:53.874 GB7TV [default INFO] Application destroyed
2021-08-09T13:21:53.874 GB7TV [default DEBUG] TmpDir deleted: buckets/tmp/process-d9194e7436d567f3
2021-08-09T13:21:53.875 GB7TV [default DEBUG] TmpDir deleted: buckets/tmp/history-6fe941bf2bfb00d4
2021-08-09T13:21:53.875 GB7TV [default DEBUG] TmpDirManager cleaning: buckets/tmp
```

This makes me more convinced that this crash happened because `WorkScheduler` may schedule some work after it's been aborted which will lead to non-empty `mChildren` which is the failure we can see in gdb.

I'll open a PR containing this proposed fix.
